File: cpp/linked_list.cpp
/// Author:  Alexander Bazzi
/// Description: Implementation of a singly linked-list

#include <iostream>
#include <array>

template<typename T>
class Node {
public:
    T data;
    Node* next;

    Node(T data, Node* next = nullptr) : data(data), next(next) {}

};

template<typename T>
class LinkedList
{
    private:
    Node<T>* head;

    public:
    explicit LinkedList() : head(nullptr) {}

    void insertEnd( T data ) {
        Node<T>* newNode = new Node<T>(data);
        if( head == nullptr ) {
            head = newNode;
        } else {
            Node<T>* current = head;
            while( current->next != nullptr ) {
               current = current->next;
            }
            current->next = newNode;
        }
    }

    Node<T>* reverseOrder() {
        Node<T>* current = head;
        Node<T>* prev = nullptr;
        Node<T>* next = nullptr;

        while(current != nullptr) {
            next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }

        head = prev;
        return head;
    }

    void printList() const {
        Node<T>* current = head;
        while(current != nullptr) {
            std::cout << current->data << std::endl;
            current = current->next;
        }
    }
};

int main()
{
    // create a linked list of ints
    LinkedList<int> list = LinkedList<int>();

    list.insertEnd(1);
    list.insertEnd(2);
    list.insertEnd(3);

    list.printList();
    list.reverseOrder();
    list.printList();
}

File: python_std/LinkedList.py

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def append(self, val): 
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node
    
    def print(self) -> None:
        current = self.head
        while current:
            print(current.val, end=" -> ")
            current = current.next
        print("None")

def main():
    ll = LinkedList()
    ll.append(32)
    ll.append(31)
    ll.append(39)
    ll.print()

if __name__ == "__main__":
    main()